packaging/CMakeLists.txt

```diff
@@ -61,7 +61,7 @@ install(TARGETS Halide_Runtime
 
 # Captures both the runtime and Halide.h
 install(DIRECTORY ${Halide_BINARY_DIR}/include/
-        TYPE INCLUDE
+        DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/halide
```

Comment:
I will not accept a patch that unconditionally adds a suffix to a GNUInstallDirs variable, sorry.

Reply:
@alexreinking this is going to be contentious then.
Can you please explain why must the headers be installed directly into `/usr/include` and not `/usr/include/halide` ?

Reply:
> Can you please explain why must the headers be installed directly into `/usr/include` and not `/usr/include/halide` ?

First, there is no "must". It is completely possible to set `-DCMAKE_INSTALL_INCLUDEDIR=include/halide` when configuring the build if that's what you need.

Hard-coding a suffix like this makes it impossible for a package maintainer for another system to remove that suffix without patching our build. It's not _always_ right to add. For instance, on Linux, libraries that are installed to `/opt` don't need it. `/opt/halide/include` is completely benign, but `/opt/halide/include/halide` is redundant at best and frustrating at worst. On Windows/Nuget, too, it's common to isolate libraries from each other like this, rather than installing them on top of each other in a common prefix.

Finally, and maybe most importantly, changing the default would be quite unexpected for our users since almost all Halide code in the wild uses `#include <Halide.h>`, not `#include <halide/Halide.h>`. I also don't see any real risk of name clashes since all of our headers are prefixed with either `Halide` or `halide_`.

Reply:
> > Can you please explain why must the headers be installed directly into `/usr/include` and not `/usr/include/halide` ?
> 
> First, there is no "must". It is completely possible to set `-DCMAKE_INSTALL_INCLUDEDIR=include/halide` when configuring the build if that's what you need.
 
> Hard-coding a suffix like this makes it impossible for a package maintainer for another system to remove that suffix without patching our build. It's not _always_ right to add. For instance, on Linux, libraries that are installed to `/opt` don't need it. `/opt/halide/include` is completely benign, but `/opt/halide/include/halide` is redundant at best and frustrating at worst. On Windows/Nuget, too, it's common to isolate libraries from each other like this, rather than installing them on top of each other in a common prefix.

Okay, i see you point.

> Finally, and maybe most importantly, changing the default would be quite unexpected for our users since almost all Halide code in the wild uses `#include <Halide.h>`, not `#include <halide/Halide.h>`. I also don't see any real risk of name clashes since all of our headers are prefixed with `Halide`.

This is a non-issue  `#include <Halide.h>` **are** correct, and will continue to work,
as per your comment: https://github.com/halide/Halide/pull/6266#discussion_r717096810

Let me reframe the question: unless overridden by user (!!!),
would it make sense to default to installing these into subdirectory?

Reply:
> This is a non-issue #include <Halide.h> are correct, and will continue to work, as per your comment: #6266 (comment)

That's right, _assuming_ the user is consuming Halide via CMake. My understanding is that our corporate users just call CMake to build a Halide distribution and then consume it with another build system (make, blaze, etc.). Changing the default will cause headaches for them. I think @abadams has more context on this.

> Let me reframe the question: unless overridden by user (!!!), would it make sense to default to installing these into subdirectory?

Because of the above, I am quite hesitant to change the default; however, it might make sense to rework our _presets_ and direct users towards those in the README / build documentation. There it might be appropriate to define a preset for a particular system that wants this convention.

Reply:
They don't even necessarily do that. E.g. Google famously compiles all third party code from scratch with their own build system. What directories headers land in in a cmake-generated install package isn't going to affect corporate users meaningfully. At most they'll have to add a -I flag somewhere to their build setup.

Reply:
Thanks, that eases my concerns somewhat. Someone out there might yet be affected, but we won't know about them until they complain, I guess. 

I still don't see the _advantage_ of making this the default, though. If you're installing to a local prefix, then you know what will be there and won't try to, like, install multiple Halide versions on top of each other. If you're creating a package, then you're most likely checking all these, anyway.

I also don't see any _harm_ if a user does put our ten or so `Halide*.h` headers in `/usr/include`. So this change just seems a bit frivolous to me.

I might change my mind if I could see how this new default would be more than aesthetics and even then worth the pain of changing the default. At least, our homebrew and vcpkg packages would need to be updated to continue conforming to the prescribed layouts, which will be a bit of a chore...

Reply:
Mostly, the same reason why docs aren't being installed into doc/ but into subdirectory.

While cmake defaults to installing docs into subdirectory
it is expected that a whole directory of headers is being installed,
and it is a common practice to name said directory after the project.

https://cmake.org/cmake/help/latest/command/install.html mentions this:
```
Projects wishing to follow the common practice of installing headers into a project-specific subdirectory will need to provide a destination rather than rely on the above.
```

The harm (at least the one i'm aware of) is subtle, and you have pointed it out already:
If the headers are not located in the directory that is in the default include path,
then unless user either includes them directly, or specifies their location via `-I`,
then they will not be found by the compiler. Having to explicitly specify the headers
that one wants to include leads to less error-prone expirience than
relying on the right ones being picked implicitly.

But as you have said, this will just work with cmake configs.

Reply:
Let me think about this more while we work on finding the right place for the autoscheduler tools.

I'm going to look through a few sets of packaging guidelines for various platforms and see if there is a common consensus on any of these issues.